### ISSUE TITLE
Post Port: Code review, refactor

### ISSUE THREAD

**Issue by hexabits** [niftools/nifskope#41]
While the actual NifSkope codebase didn't require too many changes, Qt underwent a pretty drastic overhaul between 4.8 and 5.2.  They deprecate code starting in one version, and then make it obsolete in a later version.  This all happens silently unfortunately, and you have to actually study which classes and methods have been deprecated or made obsolete.

Also, this normally shouldn't happen, but code can change behavior between versions.  For example, file filters in dialog windows stopped working.  I've already fixed it, but more strangely, I can't find any explanation as to why `*.nif` used to work, but now I need `NIF (*.nif)`.  Another example, double-clicking/dragging NIFs to open up the program stopped working because of some changes to QUrl (also fixed).

See:

[Obsolete Classes](http://qt-project.org/doc/qt-5/obsoleteclasses.html)
[Qt Coding Style](http://qt-project.org/wiki/Qt_Coding_Style)
[Qt Coding Conventions](http://qt-project.org/wiki/Coding-Conventions)
### Task List
- [x] Remove Qt4-specific code
- [x] Cross-reference deprecated/obsolete code between 4.8 and 5.2 with the NifSkope codebase.  Replace obsolete code, optionally replace deprecated code, or leave a TODO.
- [x] ~~Where there are new classes/methods available to improve or simplify the codebase, do so.~~ This was a bit too vague to be a checkable task, and is an ongoing goal. 
- [x] Use Q_UNUSED to clean up warnings.
- [x] Clean up whitespace.
- [x] Consistent code style.
- [x] Invert conditions to reduce nesting.
- [x] Remove unnecessary "if/else return" nesting.
- [x] Move from QRegExp to QRegularExpression
### Deferred Tasks
- [x] Clean up comments (formatting, spelling, grammar). _`NEEDS FUTURE TICKET`_

And also, while not a quantifiable task for the task list, the program should be tested rigorously in order to find any areas that could have broken during the port, like the aforementioned QUrl / file issues.  
### Tracking

Refactor Progress #32

**Comment by hexabits** [comment]
_(Tip: You can append ?w=1 to the url to remove purely whitespace changes)_

### Qt Support

|  |  |
| --- | --- |
| NO_CASTs to prevent using deprecated behavior | niftools/nifskope@dc217939cceae6d70c95c6ee48a38fd47c66c92c |
| Use QT_DISABLE_DEPRECATED_BEFORE 5.2 | niftools/nifskope@62778e2dd0b4c0430742deb041cace9fe04686d5 |
| Remove obsolete workaround for Qt 4.4 | niftools/nifskope@4efcd8f3906b3fd1aeb599fa16af86f9347df695 |
| Fix Qt warning when opening ColorWheel | niftools/nifskope@22611d0c62cd7c3cb975ef7f0692f6990b962b38 |
| Switch to QRegularExpression from QRegExp | niftools/nifskope@67ad1f6596b56463598b3270ac9ff254533e4193 |
| Replace remaining QRegExp | niftools/nifskope@0e480c4162ec722c6519ec2886df830a07506384 |

Disabling deprecated behavior with macros turns them into compiler errors so that they can be found and replaced. Changes were minimal. First commit involved just using the method Qt already had to return char pointers instead of using a cast. Second commit outright disabled anything deprecated as of 5.2, and the only thing that needed to be changed were the use of Qt algorithms that were deprecated in favor of the standard library.

### Build/project

|  |  |
| --- | --- |
| Clean up compiler warnings 4100 and 4189 | niftools/nifskope@43b408c7f9e617d19461aa0e903c7a93cbe71cc6 |
| Nifskope.pro C++11, comments for other devs | niftools/nifskope@1db6276945e6fb30e1bc4db8082d36242e1f7757 |
| Updated .ts paths, ran lupdate | niftools/nifskope@d981f005b941979e110ac91d863b00a31789e56c |
| Build improvements for MinGW | niftools/nifskope@4d5e1fe965900bc854f92a49712e680d6544234b |
| Additional DLLs copied for shared builds | niftools/nifskope@64cd063c1f1785fdf83c30c735af170209bca4aa |
| Disable compile/analysis warnings for qhull | niftools/nifskope@43ea2e8b2f4de705742ae477cf5448b7b80a4fce |
| Nifskope.pro C++11, comments for other devs | niftools/nifskope@1db6276945e6fb30e1bc4db8082d36242e1f7757 |

### Header/Includes

|  |  |
| --- | --- |
| Initial header cleanup | niftools/nifskope@4a59a6fae9fb023f0341a6e0f98438f86949a07b |
| Explicit includes, move to cpp where possible | niftools/nifskope@79fa8cee196a41050ade80247ceba0f7be0a07c2 |
| Remove relative includes | niftools/nifskope@e1811c9abbc635c2d1ca5e043b1976985f8d053a |
| Remove unnecessary undefs | niftools/nifskope@1a11a728797bacd0cc035700d7eaf04141885150 |
| Missing include for debug builds | niftools/nifskope@58139946e1248075d80aabf4fd40af38174f9bce |

### Whitespace

|  |  |
| --- | --- |
| Whitespace changes | niftools/nifskope@b198b5f7af2f47621eaa48d6f501547acfd0cd76 |
| Whitespace / code style changes Initial | niftools/nifskope@e9e5f36688f38714e80cdee2deab30393039be2f |

First commit is manual changes I made for the most noticeable indentation issues I happened across.  Some were using 3 spaces per indent, some were just a mix of tabs and spaces.

Second commit is automated changes using Uncrustify**. Before running it I spent literally days tweaking the settings and verifying everything went over well.  In preparation I had to add brackets to foreach macros because it's not standard C++.  It was tripping up on a few files where for example there would be a bracketed if statement inside of a braceless foreach.  I don't blame it, since it's a terrible idea to nest something bracketed under something not.  Anyway, I did manage to miss a few of these before running it, but the only thing it did was mess up the indentation level on one file (just had to Shift-Tab the code back to the first column).  There is virtually no chance that it did anything to affect the integrity of the code, as I set it to make whitespace changes only.  There are optional settings to auto-bracket single-line statements (or the opposite) and make if/else chains use consistent bracketing, but I made such changes manually in later commits.

**: I will eventually finish the code style guide and also upload the Uncrustify config file I used.

### Code Style (Formatting only)

|  |  |
| --- | --- |
| Fix up all braceless Qt foreach blocks | niftools/nifskope@7d5376879f4cacf076c027a64e6c7ba907081559 |
| More foreach brackets | niftools/nifskope@662d37ac82d27e75e01cbca1117dc9a244ac8902 |
| Remove extraneous semicolons | niftools/nifskope@67b2575332cc9a14eeb1f69801aadfe05d1d3ebb |
| Some manual cleanup, post-auto format | niftools/nifskope@e1f40b555cc5e2294a70f81f1ac9f5d2ca9fd678 |
| Consistent if/else chains, no mixed braces | niftools/nifskope@7eab8e17030f1decb352a3bddf919664322f35dd |
| For/foreach no mixing braceless with not | niftools/nifskope@499ef262e9eedc1067fe58f62c3343f2d282f0f3 |
| Misc cleanup | niftools/nifskope@61afb1224fdbd1f389305cd3b398d60e7997ed4e |
| Enum formatting | niftools/nifskope@5b789a24a4261e3e4a3ece6da8a75c89f9dfbf31 |

Aside from the addition of brackets, these changes were mostly aesthetic. The addition of brackets was to prevent future issues and improve the overall integrity of the code. There were some highly questionable if/else chains with a mix of brackets and no brackets.  There were single-line if statements separated by comments, or bracketed statements underneath braceless statements.  There were multi-nested single line statements, which is just as bad. There was one actual completely braceless if/else chain with one level of nesting (if ( if / else ) else ).  There were if/else chains that were all kept on one line per statement (no indentation).

### Code Style

|  |  |
| --- | --- |
| Reduce if/else nesting where possible | niftools/nifskope@092712d3aaf3dc4e51d6069af6713e1fa6296167 |
| Invert conditionals, reduce nesting | niftools/nifskope@1c473a9d6dbb1fac5bb6fa3879f15d367e92bc3e |
| More inverts to reduce nesting, general cleanup | niftools/nifskope@98eae83916e1fc2cd4656c6c37b409b73f152733 |
| Static code analysis fixes | niftools/nifskope@b1e86a3fc8277c1fe39163b74c31392c6df843a0 |
| Simplification with `value()`, de-nesting | niftools/nifskope@b4a38dfb3ad8ea02a2d8d6f98ca18a6c58f82980 |
| Simplify arg chains when possible | niftools/nifskope@fcf42e6bc384ee5ff84c60f5591f6d5e3916f5da |

I removed many "if return / else return" idioms.  If a scope is guaranteed to return there is no need to nest an else/else if underneath it.

I additionally inverted conditionals where possible to reduce nesting.  Take for example:

``` C++
// Unnecessary nesting

if ( true ) {
    if ( !somethingElse ) {
        // Do this
        doThis();
        doThisOtherThing();

        return true;
    } else {
        return false;
    }
} else {
    return false;
}

// Instead

if ( false )
    return false;

if ( somethingElse )
    return false;

// Do this
doThis();
doThisOtherThing();

return true;
```

### Fixes

|  |  |
| --- | --- |
| Fix a regression with getParent | jonwd7/nifskope@e8a59258727d53b710156661b22becbb7077e7e1 |
| Fix regression with kfmmodel | jonwd7/nifskope@4d1275b89b322722f7d07bc71ece83fbc1f2fc0d |
| Prevent UVWidget crash in Debug builds | niftools/nifskope@75f98f184451e10443b3be8d5b3cef34ca6daeb9 |

These undid a few changes that I made that upon review actually changed the result.  I've reviewed the changes multiple times now. 

### C++11

|  |  |
| --- | --- |
| Uniform Initialization / Initializer Lists | niftools/nifskope@cd549b0437afd6fca8f03045cf0f281f7bf119e3 |
| Uniform Initialization / Initializer Lists | niftools/nifskope@c46034a01fb46003586ad4dd2669f74c27631bd4 |
| Uniform Initialization / Initializer Lists | niftools/nifskope@1bec9bcf5f4884a5ca4183595070ac9c48fd49dc |
| Uniform Initialization / Initializer Lists | niftools/nifskope@6c7edd563fb89621f594a8a576f81e9fa2ef6a86 |
| Signal/Slot syntax 1 | niftools/nifskope@9e98b2b1edfedc9ba5f693d9ca66cebaa2f23c3c |
| Signal/Slot syntax 2 | niftools/nifskope@7d8242faba28d59bbd3606161d2005018882c677 |
| Signal/Slot syntax 3 | niftools/nifskope@154cfded4187d91cc64a4a23a958a13a0adcbfea |
| nullptr | niftools/nifskope@9436f1a8cda47de5d43ab4f91170e59cd0e661f1 |
| nullptr | niftools/nifskope@bd83b6e15b2281c9cf32b1ea775f82c0afab9141 |
| nullptr | niftools/nifskope@4e51809c295741fa8b133cfc73492280a84a5c09 |
| Range based `for` | niftools/nifskope@8375e5cd1f009b67493ecf3ead058ca644e02bd9 |
| Range based `for` | niftools/nifskope@aa9823999d3f6144e011182e856df0d099142b01 |

**Comment by hexabits** [comment]
I noticed that QRegExp is being superseded by QRegularExpression starting in Qt 5 and I started replacing it where it was used.  But I've run into some issues.  For one, I'm not sure it worked before.  The one place I've dealt with anyway, which is the node culling. 

For example if I enter in (ignoring the pipe separators) `NiTriShape` ... It does not cull NiTriShapes. If I keep with the style and use the `^` i.e. `^NiTriShape` it doesn't do anything either. I have tried (non-exhaustively):

```
Ni
^Ni
Ni*
^Ni*
NiTriShape
^NiTriShape
^NiTriShape$
NiTriShapeData
```

And so on.  And yes, Cull Node by Name is turned on.

In fact, there is another issue going on here.  The edit box should be greyed out if it is not checked, but all of my builds post-Qt5 seem to ignore it.*\*  Furthermore it's very slow to type in those fields in my builds, but the old releases are faster.

---

**It must be an intentional change in 1.1.3 or earlier.  I use 1.1.0 as a backup because opening 1.1.3 constantly resets the UI while using dev versions, and the box is greyed out in 1.1.0.  Additionally it's commented out in code:

``` C++
        CullExpr->setToolTip( tr( "Enter a regular expression. Nodes which names match the expression will be hidden" ) );
//    leave RegExp input open as both rendering and collada cull sharing this
//    CullExpr->setEnabled( CullByID->isChecked() );
        CullExpr->setEnabled( true );
        connect( CullExpr, SIGNAL( textChanged( const QString & ) ), this, SIGNAL( sigChanged() ) );
//    connect( CullByID, SIGNAL( toggled( bool ) ), CullExpr, SLOT( setEnabled( bool ) ) );
```

---

Ohhhhh.  I see the issue here. :)  I am considering the node name to be "NiTriShape" etc, when it's really the Name value on the block.   Guess that is something to add perhaps.

That still leaves the slowness with that text edit and maybe others.  I will look into all the other text edit fields.

**Comment by hexabits** [comment]
OK, I narrowed down the slowness to nothing but having Textures enabled (or loaded) across all versions.  Even 1.1.0 has sluggish UI elements once I added my texture paths back (<1.1.4 can't autodetect the Skyrim BSAs).  When you disable textures, the textbox for Cull Nodes by Name becomes "fast" again.

The presence of textures also makes the other UI elements sluggish, like the Ambient/Diffuse/Specular and Background/Foreground color wheels.

So maybe I can make improvements there so that these elements are more responsive when textures are being rendered.

---

``` C++
    connect( Options::get(), SIGNAL( sigChanged() ), textures, SLOT( flush() ) );
    connect( Options::get(), SIGNAL( sigChanged() ), this, SLOT( update() ) );
```

Narrowed it down to this in glview.cpp (old signal syntax). 

Basically, `sigChanged` might be too vague a signal.  Since almost all the options do not need a texture flush.

---

### Fix

Selectively flush textures for large speed up 5487659f2c5a08847354d175995b0eaa21373bd9

**Comment by hexabits** [comment]
Updated top comment with all basic code refactoring commits.

### Deferred Tasks
- Clean up comments (formatting, spelling, grammar).
